### PR TITLE
Fix secret name for openshift mode

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
Upgrading the gatekeeper causes cert-issue  due to the fact that the same secret name has been created by the service-ca already
 Ref: https://issues.redhat.com/browse/ACM-11413
Signed-off-by: yiraeChristineKim <yikim@redhat.com>

### CODE REVIEW

**Comment by dhaiducek** [comment]
Followup to:
- #80

File: controllers/gatekeeper_controller.go [hunk 1]
```diff
@@ -1038,7 +1039,31 @@ func openShiftDeploymentOverrides(obj *unstructured.Unstructured) error {
 
 	err = unstructured.SetNestedField(obj.Object, containers, "spec", "template", "spec", "containers")
 	if err != nil {
-		return errors.Wrapf(err, "Failed to set the OpenShift overrides")
+		return errors.Wrapf(err, "Failed to set --disable-cert-rotation for OpenShift")
```
Comment (dhaiducek) [inline controllers/gatekeeper_controller.go:1042]: This is setting all the overrides including securityContexts, not just disable cert rotation.
```suggestion
		return errors.Wrapf(err, "Failed to set the OpenShift overrides for the Deployment containers")
```

File: controllers/gatekeeper_controller.go [hunk 2]
```diff
@@ -1038,7 +1039,31 @@ func openShiftDeploymentOverrides(obj *unstructured.Unstructured) error {
...
+
+	volumes, _, err := unstructured.NestedSlice(obj.Object, "spec", "template", "spec", "volumes")
+	if err != nil || len(volumes) == 0 {
+		return errors.Wrapf(err, volumeErrMsg)
+	}
+
+	vol, ok := volumes[0].(map[string]interface{})
+	if !ok {
+		return errors.Wrapf(err, "Failed to parse volumes")
+	}
```
Comment (dhaiducek) [inline controllers/gatekeeper_controller.go:1055]: Do we need to be concerned whether there are multiple volumes here? Or is it safe to assume there's only one?

And if there's only one, we're just overwriting the name? Do we need to do some validation here? Is there a name we're expecting could be different that needs to be set?

File: controllers/gatekeeper_controller.go [hunk 3]
```diff
@@ -1038,7 +1039,31 @@ func openShiftDeploymentOverrides(obj *unstructured.Unstructured) error {
...
-		return errors.Wrapf(err, "Failed to set the OpenShift overrides")
+		return errors.Wrapf(err, "Failed to set --disable-cert-rotation for OpenShift")
+	}
+
+	const volumeErrMsg = "Failed to set the certificate volume mount for OpenShift"
+
+	volumes, _, err := unstructured.NestedSlice(obj.Object, "spec", "template", "spec", "volumes")
+	if err != nil || len(volumes) == 0 {
+		return errors.Wrapf(err, volumeErrMsg)
+	}
```
Comment (dhaiducek) [inline controllers/gatekeeper_controller.go:1050]: So we're not creating the volume if it doesn't exist?

File: test/e2e/gatekeeper_controller_test.go [hunk 4]
```diff
@@ -824,8 +832,8 @@ var _ = Describe("Gatekeeper", func() {
 					v, ok := annotations["service.beta.openshift.io/serving-cert-secret-name"]
 					g.Expect(ok).Should(BeTrue(),
 						"Should have service.beta.openshift.io/serving-cert-secret-name annotation")
-					g.Expect(v).Should(Equal("gatekeeper-webhook-server-cert"),
-						"Should be gatekeeper-webhook-server-cert")
+					g.Expect(v).Should(Equal(controllers.OpenshiftSecretName),
+						"Should be "+controllers.OpenshiftSecretName)
```
Comment (dhaiducek) [inline test/e2e/gatekeeper_controller_test.go:836]: The failure will already show the actual and expected strings, so I think a more helpful description would be what the test is about, something like:
```suggestion
					g.Expect(v).Should(Equal(controllers.OpenshiftSecretName),
						"service.beta.openshift.io/serving-cert-secret-name should have the expected value")
```

File: test/e2e/gatekeeper_controller_test.go [hunk 5]
```diff
@@ -807,6 +812,9 @@ var _ = Describe("Gatekeeper", func() {
 
 					g.Expect(webhook.Spec.Template.Spec.Containers[0].Args).
 						Should(ContainElement("--disable-cert-rotation"), "Webhook should have disabled cert arg")
+					g.Expect(webhook.Spec.Template.Spec.Volumes[0].Secret.SecretName).
+						Should(Equal(controllers.OpenshiftSecretName),
+							"Cert Secret name should be "+controllers.OpenshiftSecretName)
```
Comment (dhaiducek) [inline test/e2e/gatekeeper_controller_test.go:817]:
```suggestion
							"Webhook deployment certificate volume mount should have the expected name")
```

File: test/e2e/gatekeeper_controller_test.go [hunk 6]
```diff
@@ -795,6 +797,9 @@ var _ = Describe("Gatekeeper", func() {
 
 					g.Expect(audit.Spec.Template.Spec.Containers[0].Args).
 						Should(ContainElement("--disable-cert-rotation"), "Audit should have disabled cert arg")
+					g.Expect(audit.Spec.Template.Spec.Volumes[0].Secret.SecretName).
+						Should(Equal(controllers.OpenshiftSecretName),
+							"Cert Secret name should be "+controllers.OpenshiftSecretName)
```
Comment (dhaiducek) [inline test/e2e/gatekeeper_controller_test.go:802]:
```suggestion
							"Audit deployment certificate volume mount should have the expected name")
```

**Review by dhaiducek — CHANGES_REQUESTED** [review]
Thanks! Sorry switching branches caused some disruption. I have some additional questions/comments here.

Also, would you please update the PR description and commit to the correct Jira? The one linked looks to be the IAM removal issue.

File: controllers/gatekeeper_controller.go (hunk 3 above)
Comment (yiraeChristineKim) [reply controllers/gatekeeper_controller.go:1050]: Volumes come the cr from gatekeeper. So actually, it is 0 percent for this versions. Why is this added, maybe the next version could be was removed mistakenly. 
Should I remove?

File: controllers/gatekeeper_controller.go (hunk 2 above)
Comment (yiraeChristineKim) [reply controllers/gatekeeper_controller.go:1055]: In most of the versions of gatekeeper, these values are same.. https://github.com/stolostron/gatekeeper-operator/blob/main/pkg/bindata/bindata.go 
it comes from the upstream gatekeeper. How about we add a test to check the volumes in deployments for new future versions?

File: controllers/gatekeeper_controller.go (hunk 3 above)
Comment (dhaiducek) [reply controllers/gatekeeper_controller.go:1050]: Ah, sorry I should have looked at the deployment first. I think it's fine to leave in then. In that case we'll figure out pretty quickly if it gets removed and we can address it.

Comment (yiraeChristineKim) [reply controllers/gatekeeper_controller.go:1050]: I added tests for our next version

File: controllers/gatekeeper_controller.go (hunk 2 above)
Comment (yiraeChristineKim) [reply controllers/gatekeeper_controller.go:1055]: I added tests for our next version

Comment (dhaiducek) [reply controllers/gatekeeper_controller.go:1055]: Perhaps I'm being paranoid, but there are currently two volumes there. It seems we could parse for the `name: cert` one here?
https://github.com/open-policy-agent/gatekeeper/blame/e9d94c63b572ac3949ca2c6d60ad7c8e90e2e1f6/deploy/gatekeeper.yaml#L4910

**Review by dhaiducek — CHANGES_REQUESTED** [review]
Please update the Jira issue contained in the commit message also.

**Review by dhaiducek — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**